### ISSUE TITLE
DSL Monad

### ISSUE THREAD

**Issue by aneesharaines** [MinaProtocol/mina#5366]
Most of the pieces of the new integration test framework are in now and are being put together, but we still need the DSL monad which will wrap the control flow around error handling within the framework. #4715 contains some details about the design of the monad, but some of the details there are out of scope for the initial implementation, so I will summarize the expected behavior of the monad here.

The DSL monad is conceptually similar to the `Deferred.Or_error.t` monad, with the primary difference being that there are 2 different kinds of errors instead of one. In the DSL monad, there are "soft errors" and "hard errors".

```
type error = { hard_error: string; soft_errors: error list }
type 'a acc = { result: 'a; soft_errors: error list }
type 'a t = ('a acc, error) Deferred.Result.t
```

The semantics of the DSL are as follows: if a hard error is returned at a bind point, the computation terminates (future bind points are ignored), but if a soft error is returned at a bind point, the monad interpreter accumulates that error for later inspection and continues chaining bind points as expected. In other words, a soft error does not effect the control flow of a computation. Tests are described in this monad, and when they are executed, a test is considered successful if the result of interpreting the monad matches `Ok { soft_errors= []; _ }`. Otherwise, the test is considered a failure, and any soft errors (or a hard error, if it exists) are emitted from the test executive.

Full integration of this monad is blocked on #5380, but it will suffice to implement the monad and it's semantics, along with a simple inline test which shows that it operates as expected.

Hints for onboarding engineers:

- Code should live inside of `src/lib/integration_test_lib`. You can make a new file there called `dsl.ml`, for instance.
- To write tests for the monad, you can use the inline test ppx. The `dune runtest <library_directory> command will allow you to run tests defined in this way.
  - documentation: https://github.com/janestreet/ppx_inline_test
  - example usage: https://github.com/CodaProtocol/coda/blob/f0cc34427387f312c212b4129d2175d76d0b0249/src/lib/cache_lib/impl.ml#L277
- Janestreet's core library provides an easy way to define monads via the `Monad.Make*` functors.
  - documentation: https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Monad/index.html#module-Make
  - example usage: https://github.com/CodaProtocol/coda/blob/43cb1d654d2d83a4da04c1b9e2648c8de34c4b2e/src/lib/participating_state/participating_state.ml

**Comment by QuiteStochastic** [comment]
This is the RFC to this ticket right?

https://github.com/CodaProtocol/coda/blob/rfc/integration-tests/docs/test_dsl_spec.md

**Comment by QuiteStochastic** [comment]
merged a few weeks ago with PR https://github.com/CodaProtocol/coda/pull/5647
